src/app.js

```diff
@@ -25,107 +25,92 @@ export function app(props) {
+      (element = patch(
+        appRoot,
+        element,
+        node,
+        (node = appView(appState, appActions)),
+        (skipRender = !skipRender)
+      ))
     )
-    while ((cb = lifecycleCallbackStack.pop())) cb()
   }
 
-  function hydrate(element, map) {
-    return (
-      element &&
-      h(
-        element.tagName.toLowerCase(),
-        {},
-        map.call(element.childNodes, function(element) {
-          return element.nodeType === 3
-            ? element.nodeValue
-            : hydrate(element, map)
-        })
-      )
-    )
+  function flush(cb) {
+    while ((cb = callbackStack.pop())) cb()
+  }
+
+  function setup(module, state, actions) {
+    copy(state, module.state)
+
+    setupActions(actions, state, module.actions)
+
+    if (module.init) {
+      callbackStack.push(function() {
+        module.init(state, actions)
+      })
+    }
+
+    if (module.modules) {
+      Object.keys(module.modules).map(function(key) {
+        setup(module.modules[key], (state[key] = {}), (actions[key] = {}))
+      })
+    }
   }
 
-  function createActions(actions, withActions, lastPath) {
+  function setupActions(actions, withState, withActions) {
     Object.keys(withActions || {}).map(function(name) {
-      return typeof withActions[name] === "function"
-        ? (actions[name] = function(data) {
-            return typeof (data = withActions[name](
-              getPath(lastPath, appState),
-              getPath(lastPath, appActions),
-              data
-            )) === "function"
-              ? data(update)
-              : update(data)
-          })
-        : createActions(
-            actions[name] || (actions[name] = {}),
-            withActions[name],
-            lastPath.concat(name)
-          )
+      if (typeof withActions[name] === "function") {
+        actions[name] = function(data) {
+          return typeof (data = withActions[name](withState, actions, data)) ===
+          "function"
+            ? data(update)
+            : update(data)
+        }
+      } else {
+        setupActions(
+          (actions[name] = {}),
+          withState[name] || (withState[name] = {}),
+          withActions[name]
+        )
+      }
     })
 
-    function update(withState) {
-      if (typeof withState === "function") {
-        return update(withState(getPath(lastPath, appState)))
-      }
-      if (
-        withState &&
-        (withState = setPath(
-          lastPath,
-          merge(getPath(lastPath, appState), withState),
-          appState
-        ))
-      ) {
-        requestRender((appState = withState))
-      }
-      return appState
+    function update(data) {
+      return (
+        typeof data === "function"
+          ? update(data(withState))
+          : data && requestRender(copy(withState, data)),
+        withState
+      )
     }
   }
 
-  function set(prop, value, source) {
-    var target = merge(source)
-    target[prop] = value
+  function copy(target, source) {
```

Comment:
Does hyperapp target browsers where this cannot be `Object.assign`?

Reply:
@pspeter3 Yep, so >=IE10.